Docs: UI extensions (#386)
grafana/plugin-tools: docusaurus/docs/ui-extensions/register-an-extension.md

---
id: register-an-extension
title: Register an extension from your app plugin
sidebar_label: Register an extension
sidebar_position: 1
description: Learn how to register an extension from your app plugin.
keywords:
  - grafana
  - plugins
  - plugin
  - links
  - extensions
  - app plugins
---

## Before you begin

Be sure your plugin meets the following requirements before proceeding:

- It must be an app plugin.
- It must be preloaded by setting the [preload property](https://grafana.com/docs/grafana/latest/developers/plugins/metadata/#pluginjson) to `true` in the `plugin.json`.

> - It must be installed and *enabled*.

Reviewer: ```suggestion
- It must be installed and *enabled* for the extension to be available.
```

Its tricky as this doc is about creating the extension logic in the first place, but this point in particular is more about "your extensions will only be activated if this is true".

Reviewer: I thought it would be good to link to someone on enabling app plugins, but I think we don't actually document this anywhere - I certainly can't find it. Maybe something for us to follow up on in https://grafana.com/docs/grafana/latest/administration/plugin-management/ @josmperez

> - It must be installed and **enabled**.

Author: Thank you 🙏🏻 Do we have any guidelines that I can read for these kind of conventions?

> ## How to add an extension to Grafana core/plugins from your app plugin

Reviewer: ```suggestion
## How to extend the Grafana UI or an app plugin from your app plugin
```

Not sure we should have `/` in a section header. Open to other suggestions on title @josmperez

> 1. Open the `module.ts(x)` file in your plugin where your `AppPlugin` should be exposed.

Reviewer: Is this step required? 

" in your plugin where your `AppPlugin` should be exposed." feels like overkill. You'd only have one module.ts right?

Author: Agree 👍🏻

> 1. Use the `configureExtensionLink` on your `AppPlugin` instance to register extensions. This method requires an object containing the following properties:

Reviewer: ```suggestion
1. Add the `configureExtensionLink` method in your `AppPlugin` instance to register extensions. This requires an object containing the following properties:
```

>    - `extensionPointId` *required* - the unique identifier of the extension point you would like to extend.

Author: Good idea!

   - `title` *required* - used to display your extension at the extension point.
   - `description` *required* - short description of what your extension does.

>    - `path` - a path within your app plugin where you would like to navigate the user when he/she click the extension.

Reviewer 2: ```suggestion
   - `path` - a path within your app plugin where you would like to send users when they click the extension.
```

>    - `onClick` - a callback that should be triggered when the user click the extension.

Reviewer 2: ```suggestion
   - `onClick` - a callback that should be triggered when the user clicks the extension.
```

   - `category` - a category that we should use to group your extension with other extensions.
   - `icon` - an icon that should be used while displaying your extension.

>    - `configure` - a function that will be called prior to displaying the extension which enables you to dynamically change/hide your extension depending on the context.

Reviewer 2: ```suggestion
   - `configure` - a function that is called prior to displaying the extension which enables you to dynamically change or hide your extension depending on its context.
```

::: note

> One of `path` or `onClick` is required. Otherwise the extension will be hidden due to incorrect configuration.

Reviewer 2: ```suggestion
Use either `path` or `onClick` (one is required). Otherwise, the extension will be hidden.
```

:::

> ### Example: Add a regular link

Author: I think that makes sense! Good suggestion.

> ### Example: Adding a regular link

Reviewer 2: ```suggestion
### Example: Add a regular link
```

> In the example below we are adding an extension link to the Grafana dashboard panel menu. When the user clicks the menu item "Go to basic app" he/she will be navigated to `/a/myorg-basic-app/one`.

Reviewer 2: ```suggestion
In the following example, we add an extension link to the Grafana dashboard panel menu. When the user clicks "Go to basic app," they are sent to `/a/myorg-basic-app/one`.
```

```typescript
new AppPlugin().configureExtensionLink({
  title: 'Go to basic app',
  description: 'Will send the user to the basic app',
  extensionPointId: 'grafana/dashboard/panel/menu',
  path: '/a/myorg-basic-app/one', // Must start with "/a/<PLUGIN_ID>/"
});
```

> ### Example: Adding a regular link with query string parameters from the context

Reviewer 2: ```suggestion
### Example: Add a regular link with query string parameters from the context
```

> In the example below we are adding an extension link to the Grafana dashboard panel menu. When the user click the menu item "Go to basic app" he/she will be navigated to `/a/myorg-basic-app/one?panelId=12345&timeZone=utc`.

Reviewer 2: ```suggestion
In the following example, we add an extension link to the Grafana dashboard panel menu. When the user clicks "Go to basic app," they are sent to `/a/myorg-basic-app/one?panelId=12345&timeZone=utc`.
```

```typescript
new AppPlugin().configureExtensionLink({
  title: 'Go to basic app',
  description: 'Will send the user to the basic app',
  extensionPointId: 'grafana/dashboard/panel/menu',
  path: '/a/myorg-basic-app/one', // Must start with "/a/<PLUGIN_ID>/"
  configure: (context: PanelContext) => {
    const {timeZone, panelId } = context;

    // You only need to return the properties that you would like to override.
    return {
      path: `/a/myorg-basic-app/one?panelId=${panelId}&timeZone=${timeZone}`
    };
  }
});
```

> ### Example: Adding a regular link that should be visible depending on the context

Reviewer 2: ```suggestion
### Example: Add a regular link that should be visible depending on the context
```

> In the example below we are adding an extension link to the Grafana dashboard panel menu. It will only be visible for panels with time zone set to UTC.

Reviewer 2: ```suggestion
In the following example, we add an extension link to the Grafana dashboard panel menu. It will only be visible for panels with the time zone set to UTC.
```

```typescript
new AppPlugin().configureExtensionLink({
  title: 'Go to basic app',
  description: 'Will send the user to the basic app',
  extensionPointId: 'grafana/dashboard/panel/menu',
  path: '/a/myorg-basic-app/one', // Must start with "/a/<PLUGIN_ID>/"
  configure: (context: PanelContext) => {
    const {timeZone } = context;

    switch (toLowerCase(timeZone)) {
      case 'utc':
        return {}; // no overrides applied but we want to display the extension.
      default:
        return undefined; // returning undefined from the configure function will hide the extension.
    }
  }
});
```

> ### Example: Adding a link that should display a modal

Reviewer 2: ```suggestion
### Example: Add a link that should display a modal
```

> In the example below we are adding an extension link to the Grafana dashboard panel menu. It will open a flow (defined in our app) in a modal on top of the current view.

Reviewer 2: ```suggestion
In the following example, we add an extension link to the Grafana dashboard panel menu. It will open a flow (defined in our app) in a modal on top of the current view.
```
